### PR TITLE
feat: implementation of new SymbolicIndexingInterface

### PR DESCRIPTION
**Pull request by AayushSabharwal**

### CODE REVIEW

File: src/systems/abstractsystem.jl [hunk 1]
```diff
@@ -174,14 +182,103 @@ function SymbolicIndexingInterface.independent_variables(sys::AbstractSystem)
     end
 end
 
-function SymbolicIndexingInterface.independent_variables(sys::AbstractTimeDependentSystem)
-    [getfield(sys, :iv)]
+#Treat the result as a vector of symbols always
+function SymbolicIndexingInterface.is_variable(sys::AbstractSystem, sym)
+    if unwrap(sym) isa Int    # [x, 1] coerces 1 to a Num
```
Comment (oscardssmith) [inline src/systems/abstractsystem.jl:187]: this seems really ugly.

Comment (AayushSabharwal) [reply src/systems/abstractsystem.jl:187]: Yeah it is. It's the only way I could find to cover all the cases

Comment (oscardssmith) [reply src/systems/abstractsystem.jl:187]: what does `unwrap` do?

Comment (AayushSabharwal) [reply src/systems/abstractsystem.jl:187]: It gets the value inside a `Num`, which is usually either a `SymbolicUtils.BasicSymbolic` for symbolic variables or a number in the case of something like `Num(1)`

File: src/systems/abstractsystem.jl [hunk 2]
```diff
@@ -224,8 +224,12 @@ function SymbolicIndexingInterface.variable_symbols(sys::AbstractSystem)
 end
 
 function SymbolicIndexingInterface.is_parameter(sys::AbstractSystem, sym)
+    usym = unwrap(sym)
     if unwrap(sym) isa Int
-        return unwrap(sym) in 1:length(parameters(sys))
+        return usym in 1:length(parameters(sys))
+    end
+    if istree(usym) && operation(usym) != getindex
```
Comment (oscardssmith) [inline src/systems/abstractsystem.jl:231]: `return !istree(usym) || operation(usym) == getindex`

**Pull request merged**